src/CodeFormatter.cs

```diff
@@ -293,99 +289,105 @@ private static void LogWorkspaceDiagnostics(ILogger logger, bool logWorkspaceWar
+            var configSet = await GetAnalyzerConfigSetAsync(document.Project, cancellationToken).ConfigureAwait(false);
+            var result = configSet.GetOptionsForSourcePath(document.FilePath!);
 
-            if (document is null || await ShouldIgnoreDocument(document, fileMatcher, includeGeneratedFiles, cancellationToken))
+            if (result.AnalyzerOptions.Count == 0)
             {
-                return (null, null, null, false);
+                return null;
             }
 
-            var context = await codingConventionsManager.GetConventionContextAsync(
-                document.FilePath, cancellationToken).ConfigureAwait(false);
-
-            OptionSet options = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
-
-            // Check whether an .editorconfig was found for this document.
-            if (context?.CurrentConventions is null)
+            // The compiler will reuse the same instance of AnalyzerOptions if matched .editorconfig sections
+            // for multiple files are the same.
+            if (s_editorConfigOptions.TryGetValue(result, out var editorConfigOptions))
             {
-                return (document, options, null, false);
+                return editorConfigOptions;
             }
 
-            options = optionsApplier.ApplyConventions(options, context.CurrentConventions, project.Language);
-            return (document, options, context.CurrentConventions, true);
+            var configOptions = new EditorConfigOptions(result.AnalyzerOptions);
+            s_editorConfigOptions[result] = configOptions;
+            return configOptions;
         }
 
-        private static async Task<bool> ShouldIgnoreDocument(
-            Document document,
-            Matcher fileMatcher,
-            bool includeGeneratedFiles,
-            CancellationToken cancellationToken)
+        private static async Task<AnalyzerConfigSet> GetAnalyzerConfigSetAsync(Project project, CancellationToken cancellationToken)
         {
-            if (!fileMatcher.Match(document.FilePath).HasMatches)
-            {
-                // If a files list was passed in, then ignore files not present in the list.
-                return true;
-            }
-            else if (!document.SupportsSyntaxTree)
+            if (s_analyzerConfigSets.TryGetValue(project.Id, out var configSet))
             {
-                return true;
+                return configSet;
             }
-            else if (!includeGeneratedFiles && await GeneratedCodeUtilities.IsGeneratedCodeAsync(document, cancellationToken).ConfigureAwait(false))
-            {
-                // Ignore generated code files.
-                return true;
-            }
-            else
+
+            var parseConfigs = project.AnalyzerConfigDocuments.Select(ParseAnalyzerConfigAsync);
+            var configs = await Task.WhenAll(parseConfigs);
+            configSet = AnalyzerConfigSet.Create(configs);
```

Comment:
Do we need to expose a public API to get the workspace-produced AnalyzerConfigSet? If that's internal we should just make it public, or make a method on Document that returns the document-specific AnalyzerConfig.